### PR TITLE
Check labels in e2e PipelineRun tests and document label propagation

### PR DESCRIPTION
**Pull request by dwnusbaum**
The e2e tests in test/pipelinerun.go now assert that custom labels set on the PipelineRun are propagated through to the TaskRun and Pod, and that the static labels that are added to all TaskRuns and Pods are propagated correctly as well.

In addition, documentation has been added to explain that labels are propagated and to mention the specific labels that are added automatically to generated TaskRuns and Pods.

This is a follow-up to #488.

CC @shashwathi @bobcatfish

### CODE REVIEW

File: docs/pipelineruns.md [hunk 1]
```diff
@@ -97,6 +97,26 @@ of the `TaskRun` resource object.
 For examples and more information about specifying service accounts, see the
 [`ServiceAccount`](./auth.md) reference topic.
 
+## Labels
```
Comment (dwnusbaum) [inline docs/pipelineruns.md]: This seemed like the least bad place for the docs, but I don't know if it really fits. Any suggestions are welcome!

File: test/pipelinerun_test.go [hunk 2]
```diff
@@ -322,10 +345,12 @@ func getPipelineRunSecret(suffix int, namespace string) *corev1.Secret {
 }
 
 func getHelloWorldPipelineRun(suffix int, namespace string) *v1alpha1.PipelineRun {
-	return tb.PipelineRun(getName(pipelineRunName, suffix), namespace, tb.PipelineRunSpec(
-		getName(pipelineName, suffix),
-		tb.PipelineRunServiceAccount(fmt.Sprintf("%s%d", saName, suffix)),
-	))
+	return tb.PipelineRun(getName(pipelineRunName, suffix), namespace,
+		tb.PipelineRunLabel("hello-world-key", "hello-world-value"),
```
Comment (dwnusbaum) [inline test/pipelinerun_test.go]: This just adds a custom label to the `service account propagation` test. @shashwathi is that what you had in mind, or do you think I should create a separate test for labels specifically? (I wasn't sure how conservative we wanted to be about adding new e2e tests)

**Comment by bobcatfish** [comment]
/ok-to-test

(Still waiting to hear back on adding you to the org @dwnusbaum , sorry for the delay!!)

File: docs/pipelineruns.md (hunk 1 above)
Comment (bobcatfish) [reply docs/pipelineruns.md]: i had trouble thinking of a good place myself! the only other alternative i can think of is to create a new doc like `troubleshooting.md` or `faq.md` or `tips.md` or something but none of those options seem great either 🤔 this seems fine for now tho if you don't like any of those ideas!

File: docs/pipelineruns.md [hunk 3]
```diff
@@ -97,6 +97,26 @@ of the `TaskRun` resource object.
...
+* `pipeline.knative.dev/pipelineRun` will contain the name of the `PipelineRun`
+
+These labels make it easier to find the resources that are associated with a
+given pipeline.
+
+For example, to find all `Pods` created by a `Pipeline` named test-pipeline,
+you could use the following command:
+
+```shell
+kubectl get pods --all-namespaces -l pipeline.knative.dev/pipeline=test-pipeline
```
Comment (bobcatfish) [inline docs/pipelineruns.md]: niiiice

File: test/pipelinerun_test.go [hunk 4]
```diff
@@ -160,14 +161,36 @@ func TestPipelineRun(t *testing.T) {
...
+				if err != nil {
+					t.Fatalf("Couldn't get expected Pod for %s: %s", taskRunName, err)
+				}
+				if numPods := len(pods.Items); numPods != 1 {
+					t.Fatalf("Expected 1 Pod for %s, but got %d Pods", taskRunName, numPods)
+				}
+				pod := pods.Items[0]
+				for key, taskRunVal := range labels {
+					if podVal := pod.ObjectMeta.Labels[key]; podVal != taskRunVal {
+						t.Errorf("Expected label %s=%s but got %s", key, taskRunVal, podVal)
```
Comment (bobcatfish) [inline test/pipelinerun_test.go]: total nitpick: could these be in a separate function called by the test? for the same of readability (test is getting long!)

File: test/pipelinerun_test.go [hunk 5]
```diff
@@ -160,14 +161,36 @@ func TestPipelineRun(t *testing.T) {
...
-					pipelineRunName: pipeline.PipelineRunLabelKey,
-					pipelineName:    pipeline.PipelineLabelKey,
-				} {
-					expectedName := getName(name, i)
-					lbl := pipeline.GroupName + key
-					if val := r.Labels[lbl]; val != expectedName {
-						t.Errorf("Expected label %s=%s but got %s", lbl, expectedName, val)
+
+				// Check label propagation to Tasks
+				labels := make(map[string]string, len(pr.ObjectMeta.Labels)+3)
```
Comment (bobcatfish) [inline test/pipelinerun_test.go]: why +3 here? (maybe explain with a comment or a `const` ?)

File: test/pipelinerun_test.go [hunk 6]
```diff
@@ -160,14 +161,36 @@ func TestPipelineRun(t *testing.T) {
...
+				for key, val := range pr.ObjectMeta.Labels {
+					labels[key] = val
+				}
+				labels[pipeline.GroupName+pipeline.PipelineLabelKey] = getName(pipelineName, i)
+				labels[pipeline.GroupName+pipeline.PipelineRunLabelKey] = getName(pipelineRunName, i)
```
Comment (bobcatfish) [inline test/pipelinerun_test.go]: what is the significance of `pipeline.GroupName+pipeline.PipelineRunLabelKey` and `pipeline.GroupName+pipeline.PipelineLabelKey`? (i assume these correspond to some label generation logic in the controller - maybe a comment here would help?)

File: test/pipelinerun_test.go (hunk 4 above)
Comment (dwnusbaum) [reply test/pipelinerun_test.go]: Yeah good point, I'll separate it out!

File: test/pipelinerun_test.go (hunk 6 above)
Comment (dwnusbaum) [reply test/pipelinerun_test.go]: Yeah those are static labels added by the controller, but definitely could use a comment.

File: test/pipelinerun_test.go (hunk 5 above)
Comment (dwnusbaum) [reply test/pipelinerun_test.go]: `len(pr.ObjectMeta.Labels)` for custom labels, `+2` for the static pipeline/pipelineRun label from the PipelineRun controller, and `+1` for the taskRun label from the `TaskRun` controller. I'll add a comment 😄

File: docs/pipelineruns.md (hunk 1 above)
Comment (dwnusbaum) [reply docs/pipelineruns.md]: Yeah, it seems kind of weird to create a new file just for this information. My preference would be to leave it for now, and then if we end up with other troubleshooting-esque docs we can refactor them into a single place later. Does that seem ok?

**Comment by dwnusbaum** [comment]
/retest since CI failure appears to be infra related:

>  ERROR: (gcloud.beta.container.clusters.create) ResponseError: code=400, message=No valid versions with the prefix "1.11.7" found.

File: test/pipelinerun_test.go (hunk 4 above)
Comment (dwnusbaum) [reply test/pipelinerun_test.go]: Did some refactoring in 37ed488, let me know if that seems good or if you'd like to see a totally separate `assertLabels` function with all the label-related stuff to simplify further.

File: test/pipelinerun_test.go (hunk 2 above)
Comment (shashwathi) [reply test/pipelinerun_test.go]: This looks good. Thank you @dwnusbaum

**Pull request merged**